zilliqa/src/consensus.rs

```diff
@@ -852,64 +898,55 @@ impl Consensus {
 
         self.votes.insert(
             block_hash,
-            (signatures, cosigned, cosigned_weight, supermajority_reached),
+            (
+                signatures,
+                cosigned.clone(),
+                cosigned_weight,
+                supermajority_reached,
+            ),
         );
 
         Ok(None)
     }
 
-    fn get_next_committee(&mut self, mut committee: Committee) -> Committee {
-        if committee.is_empty() {
-            panic!("committee is empty, this should never happen");
-        }
-
-        if !self.pending_peers.is_empty() {
-            info!(
-                "*** adding {} pending peers to committee",
-                self.pending_peers.len()
-            );
-        }
-
-        committee.add_validators(self.pending_peers.drain(..));
-        committee
-    }
-
     fn are_we_leader_for_view(&mut self, parent_hash: Hash, view: u64) -> bool {
         match self.leader_for_view(parent_hash, view) {
-            Some(leader) => leader == self.peer_id(),
+            Some(leader) => leader == self.public_key(),
             None => false,
         }
     }
 
-    fn leader_for_view(&mut self, parent_hash: Hash, view: u64) -> Option<PeerId> {
-        let parent = self.get_block(&parent_hash);
-
-        let parent = match parent {
-            Ok(Some(parent)) => parent,
-            _ => {
+    fn leader_for_view(&mut self, parent_hash: Hash, view: u64) -> Option<NodePublicKey> {
+        if let Ok(Some(parent)) = self.get_block(&parent_hash) {
+            let leader = self.leader_at_block(&parent, view);
+            Some(leader.public_key)
+        } else {
+            if view > 1 {
                 warn!(
                     "parent not found while determining leader for view {}",
                     view
                 );
                 return None;
             }
+            let leader = self.leader(view);
+            Some(leader.public_key)
+        }
+    }
+
+    fn committee_for_hash(&mut self, parent_hash: Hash) -> Result<Vec<NodePublicKey>> {
+        let Ok(Some(parent)) = self.get_block(&parent_hash) else {
+            return Err(anyhow!("parent block not found: {:?}", parent_hash));
         };
 
-        Some(self.leader(&parent.committee, view).peer_id)
-    }
+        let curr_root_hash = self.state.root_hash()?;
 
-    fn committee_for_hash(&mut self, parent_hash: Hash) -> Result<Committee> {
-        let parent = self.get_block(&parent_hash);
+        self.state.set_to_root(H256(parent.state_root_hash().0));
 
-        let parent = match parent {
-            Ok(Some(parent)) => parent,
-            _ => {
-                warn!("parent not found during committee_for_hash");
-                return Err(anyhow!("parent not found during committee_for_hash"));
-            }
-        };
+        let committee = self.state.get_stakers()?;
+
+        self.state.set_to_root(H256(curr_root_hash.0));
```

Comment:
Requiring `&mut self` can be avoided by using `state.at_root()` rather than `state.set_to_root()`. Keeping it const should allow eliminating `mut` in other places up the chain and I think is definitely worthwhile.